File: server/shared/dotnet/result/D2Result.cs
// -----------------------------------------------------------------------
// <copyright file="D2Result.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.Result;

using System.Net;
using D2.Shared.I18n;

/// <summary>
/// Represents the result of an operation, including success status, messages, errors,
/// and related metadata. The non-generic base type for results that do not carry a payload.
/// </summary>
/// <remarks>
/// <para>
/// <see cref="Messages"/> entries are <see cref="TKMessage"/> instances — translation
/// keys with optional parameter bindings. The wire format ships them as
/// <c>{ "key": "..." }</c> objects; SvelteKit / browser-side Paraglide translates
/// them on receipt. Server-side translation (for outbound notifications) goes
/// through <see cref="ITranslator"/>.
/// </para>
/// <para>
/// Producers obtain <see cref="TKMessage"/> instances exclusively via the
/// SrcGen-emitted <c>TK</c> constants (e.g. <c>TK.Common.Errors.NOT_FOUND</c>);
/// the type system makes "untranslated literal in <c>Messages</c>" structurally
/// unrepresentable.
/// </para>
/// </remarks>
public partial class D2Result
{
    /// <summary>
    /// Initializes a new instance of the <see cref="D2Result"/> class.
    /// </summary>
    /// <param name="success">
    /// Whether the operation was successful. Required.
    /// </param>
    /// <param name="messages">
    /// Translation messages related to the operation. Optional; defaults to empty.
    /// </param>
    /// <param name="inputErrors">
    /// Per-field input validation errors. Optional; defaults to empty.
    /// </param>
    /// <param name="statusCode">
    /// The <see cref="HttpStatusCode"/> for the operation. Optional; defaults to
    /// <see cref="HttpStatusCode.OK"/> on success and <see cref="HttpStatusCode.BadRequest"/>
    /// on failure.
    /// </param>
    /// <param name="errorCode">
    /// A standardized error code for known failure conditions. Optional.
    /// </param>
    /// <param name="traceId">
    /// Trace identifier for correlating logs and diagnostics. Optional.
    /// </param>
    public D2Result(
        bool success,
        IReadOnlyList<TKMessage>? messages = null,
        IReadOnlyList<InputError>? inputErrors = null,
        HttpStatusCode? statusCode = null,
        string? errorCode = null,
        string? traceId = null)
    {
        Success = success;
        Messages = messages ?? [];
        InputErrors = inputErrors ?? [];
        StatusCode = statusCode ?? (success ? HttpStatusCode.OK : HttpStatusCode.BadRequest);
        ErrorCode = errorCode;
        TraceId = traceId;
    }

    /// <summary>
    /// Gets a value indicating whether the operation was successful.
    /// </summary>
    public bool Success { get; }

    /// <summary>
    /// Gets a value indicating whether the operation failed.
    /// </summary>
    public bool Failed => !Success;

    /// <summary>
    /// Gets the translation messages related to the operation. Each message is a
    /// <see cref="TKMessage"/> (translation key + optional parameter bindings).
    /// </summary>
    public IReadOnlyList<TKMessage> Messages { get; }

    /// <summary>
    /// Gets the per-field input validation errors. Each <see cref="InputError"/>
    /// pairs a field name with one or more <see cref="TKMessage"/> entries
    /// describing what's wrong with that field.
    /// </summary>
    public IReadOnlyList<InputError> InputErrors { get; }

    /// <summary>
    /// Gets the <see cref="HttpStatusCode"/> for the operation.
    /// </summary>
    public HttpStatusCode StatusCode { get; }

    /// <summary>
    /// Gets the standardized error code, if applicable.
    /// </summary>
    public string? ErrorCode { get; }

    /// <summary>
    /// Gets the trace identifier for correlating logs and diagnostics, if available.
    /// </summary>
    public string? TraceId { get; }

    /// <summary>
    /// Returns a new <see cref="D2Result"/> with the same shape (Success,
    /// Messages, InputErrors, StatusCode, ErrorCode) but with the supplied
    /// <paramref name="traceId"/> in place of the original. Used by
    /// <c>BaseHandler.RunCorePipelineAsync</c> to auto-inject the request
    /// trace id on every result that crosses the handler boundary, so
    /// handlers don't have to thread it through every <c>D2Result.Ok(...)</c>
    /// call site.
    /// </summary>
    /// <param name="traceId">The trace id to attach.</param>
    /// <returns>A new <see cref="D2Result"/> with <paramref name="traceId"/> applied.</returns>
    public D2Result WithTraceId(string? traceId)
        => new(Success, Messages, InputErrors, StatusCode, ErrorCode, traceId);
}

File: server/shared/typescript/result/src/d2-result.ts
// -----------------------------------------------------------------------
// Copyright (c) DCSV. All rights reserved.
// -----------------------------------------------------------------------

import { HttpStatusCode } from "./http-status-codes.js";
import type { InputError } from "./input-error.js";
import type { ErrorCategory } from "@d2/error-category";
import type { TKMessage } from "@d2/i18n-abstractions";

/**
 * Constructor input for `D2Result`. All optional except `success`.
 */
export interface D2ResultInit<T = void> {
  readonly success: boolean;
  readonly data?: T;
  readonly messages?: readonly TKMessage[];
  readonly inputErrors?: readonly InputError[];
  readonly statusCode?: HttpStatusCode;
  readonly errorCode?: string;
  readonly traceId?: string;
  readonly category?: ErrorCategory;
}

/**
 * Result of an operation. Mirrors .NET `D2.Shared.Result.D2Result` /
 * `D2Result<T>` — same field shape so cross-language wire is byte-identical.
 *
 * Producers should always prefer a semantic factory
 * (`D2Result.notFound()`, `D2Result.unauthorized()`, etc.) over raw `fail()`
 * when one matches the failure mode — semantic factories carry the canonical
 * status code, error code, and default message together.
 */
export class D2Result<T = void> {
  readonly success: boolean;
  readonly data: T | undefined;
  readonly messages: readonly TKMessage[];
  readonly inputErrors: readonly InputError[];
  readonly statusCode: HttpStatusCode;
  readonly errorCode: string | undefined;
  readonly traceId: string | undefined;
  readonly category: ErrorCategory | undefined;

  constructor(init: D2ResultInit<T>) {
    this.success = init.success;
    this.data = init.data;
    this.messages = init.messages ?? [];
    this.inputErrors = init.inputErrors ?? [];
    this.statusCode =
      init.statusCode ??
      (init.success ? HttpStatusCode.OK : HttpStatusCode.BadRequest);
    this.errorCode = init.errorCode;
    this.traceId = init.traceId;
    this.category = init.category;
  }

  /** True when the operation failed. */
  get failed(): boolean {
    return !this.success;
  }

  /** True when the result represents partial success (HTTP 206 / SOME_FOUND). */
  get isPartialSuccess(): boolean {
    return this.statusCode === HttpStatusCode.PartialContent;
  }

  /**
   * Returns a new `D2Result` with the same shape but `traceId` swapped.
   * Used by the handler pipeline to auto-inject the request trace id on
   * results that cross the handler boundary.
   */
  withTraceId(traceId: string | undefined): D2Result<T> {
    return new D2Result<T>({
      success: this.success,
      data: this.data,
      messages: this.messages,
      inputErrors: this.inputErrors,
      statusCode: this.statusCode,
      errorCode: this.errorCode,
      traceId,
      category: this.category,
    });
  }
}
